### ISSUE TITLE
TypedAdapterFactory is not loaded with ServiceLoader on Android

### ISSUE THREAD

**Issue by almozavr** [immutables/immutables#100]
`TypedAdapterFactory` seems like is not generated (or included) into final `app.apk/META-INF/services`.
Could be an android build tools bug, and here is a [workaround](https://code.google.com/p/android/issues/detail?id=59658#c22)

Also, those are normally generated and present:
`javax.ws.rs.ext.MessageBodyReader`
`javax.ws.rs.ext.MessageBodyWriter`

build.gradle is configured like this:
```
provided 'org.immutables:value:2.0.6'
apt 'org.immutables:value:2.0.6'
compile 'org.immutables:gson:2.0.6'
apt 'org.immutables:gson:2.0.6'
```
TypeAdapters are generated as well and I can link 'em manually.

**Comment by elucash** [comment]
I have to find out what it does in Android studio...
Thanks!

BTW (org.immutables:gson artifact doesn't have processors, so "apt" scope is auxiliary for it)

**Comment by elucash** [comment]
Tried on a latest Android Studio. We have services files generated in classpath (`app/build/intermediates/classes/release/META-INF/services/*`) and it works when launching from Java. However, as was pointed out in the original description, they are not being copied into `*.apk`. I don't know how to copy them. Can some Gradle hook could do this? Currently I see no clear way to generate files during the annotation processing so they will end up in the place, from where they will be auto-copied to APK. 

Currently, I would suggest to manually register factory classes with `GsonBuilder` or manually copy generated services files to the `main/resources/*` (and keep them in sync, which is more complicated). But solution might for this might appear.

**Comment by almozavr** [comment]
Thanks for your research, let's hope android plugin will have a solution for this one day ) `Gradle` hook seems to have some prospect.

**Comment by almozavr** [comment]
Seems like the problem is not here: I'm using `retrolambda` and there is something wrong with output dir gson adapter factory puts service in:
![](http://monosnap.com/image/8NjuqYcN0Ndn1bDCkFDCSHZMXICKTf.png)

**Comment by almozavr** [comment]
https://github.com/evant/gradle-retrolambda/issues/110

**Comment by elucash** [comment]
I hope it could be addressed by `retrolambda`, but I've tested without it and it those classes still were not put inside apk for some reason.
![android-studio-sample-app](https://cloud.githubusercontent.com/assets/1692611/7542975/df97d94e-f575-11e4-9c0e-146da9ef5426.png)

**Comment by almozavr** [comment]
As retrolambda provided fix for evant/gradle-retrolambda#110 it's now clear, that the problem is in android gradle plugin :(

**Comment by almozavr** [comment]
Also, it's interesting, that `GsonMessageBodyProvider` generates it's `meta-inf/services` member that finds its way to apk. Like 
![](http://monosnap.com/image/ZFRt2VyLYKYRRY0RIlMNWNTBY8DzxQ.png)

Could you think of why does it happen? I see @Metainf.Service notation, could it be linked with possible solution?

**Comment by elucash** [comment]
It's easy explainable, the service files you see are packaged because they are just copied from prebuilt jar file. Those files are not generated during compilation of android application but was generated during compilation of Immutables jars which are available on Central. I guess if you use `@Metainf.Service` in android build, you would end with the same issue.

**Comment by almozavr** [comment]
You're right. I'm sorry, I used dep. as `compile` versus `provided`, so services bundled into the jar leaked into the apk.

**Comment by elucash** [comment]
JFYI `org.immutables:gson` module has dual intent. It can be used for compilation only. Or as compile/runtime dependency if you need any of additional features described in [JSON guide](http://immutables.github.io/json.html#gson)

**Comment by imminent** [comment]
Is there a resolution for this in configuration? Right now I have to manually copy the file to get it to work.

**Comment by elucash** [comment]
There is no resolution as I know of. For Android, the better way to load type adapters is to generate per package factories and registering newly instantiated instances with `GsonBuilder`. Please see  #287 and  #230

**Comment by imminent** [comment]
Oh, thanks @elucash makes sense.

**Comment by vovkab** [comment]
Not sure if this similar problem, but right now I can't even compile most of the time:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileRetrolambdaDebug'.
> java.io.FileNotFoundException: my-project/build/intermediates/classes/project/debug/META-INF/services
```

I'm registering all type adapters manually, so I don't really need services, is there any good way to completely disable generation or maybe there is another way to fix it?

immutables: 2.2 (the same on 2.19)
android gradle plugin: 2.1.0
gradle: 2.10

**Comment by elucash** [comment]
Unfortunately I cannot say anything about how gradle/retrolambda/javac work together and what output directories should be. But I absolutely agree that this should not prevent compilation if not used.

**Comment by vovkab** [comment]
@elucash do you know if there any way to disable service file generation?

**Comment by elucash** [comment]
I'm planning to add some knob to disable META-INF for Gson and also make warning instead of compilation error (if possible on the side of Immutables). Can you, please, copy paste the bug report as a separate issue to track?

**Comment by zsavely** [comment]
I recently stumbled upon a similar issue and created a workaround that would copy the initial generated `com.google.gson.TypeAdapterFactory` into the `app\src\main\resources\META-INF\services`. In this way, `ServiceLoader` will load implementations successfully.

Just paste this to your `build.gradle`. 
```groovy
afterEvaluate {
  android.applicationVariants.all { variant ->
    variant.javaCompiler.dependsOn(copyAdapterFactories)
  }
}

task copyAdapterFactories(type: Copy) {
  from 'build/intermediates/classes/debug/META-INF/services/com.google.gson.TypeAdapterFactory'
  into { 'src/main/resources/META-INF/services' }
}
```

After you have pasted the snippet, you can use `ServiceLoader` to include all `TypeAdapterFactory`s generated by the Immutables library.

```java
import com.google.gson.GsonBuilder;
import java.util.ServiceLoader;

GsonBuilder gsonBuilder = new GsonBuilder();
for (TypeAdapterFactory factory : ServiceLoader.load(TypeAdapterFactory.class)) {
  gsonBuilder.registerTypeAdapterFactory(factory);
}
```

**Comment by HodGreeley** [comment]
FYI it appears there was a bug in which META-INF subdirectories were being explicitly excluded.  See [this issue thread](https://issuetracker.google.com/u/1/issues/36978763).  This appears to have been fixed.  I'm currently getting META-INF correctly included when placed under main/resources.  Not sure if /resources is the only path that works.

**Comment by almozavr** [comment]
We've created simple annotation processor to aggregate all `TypeAdapterFactory` classes generated by Immutables.org - https://github.com/techery/immutables-gson-adapters-provider

The reason why we dumped `ServiceLoader` is it mystically fails on some kitkat devices and completely fails on Amazon Testing Factory (as they re-sign apk and meta-inf/services  is ripped of during this process).
